File: sol-py/smoke-test/sol.py
import socket
import threading

def handle_client(conn, addr):
    """
    Handles a single client connection.
    Args:
        conn: The socket object for the connected client.
        addr: The IP address and port of the client.
    """
    print(f"[NEW CONNECTION] {addr} connected.")
    
    try:
        while True:
            data = conn.recv(4096)
            
            # If recv returns an empty bytes object (b''), the client has 
            # shut down their sending side.
            if not data:
                break
            
            # sendall ensures the entire buffer is sent, handling network 
            # buffers automatically.
            conn.sendall(data)
            
    except ConnectionResetError:
        print(f"[ERROR] Connection forcibly closed by {addr}")
    except Exception as e:
        print(f"[ERROR] An unexpected error occurred with {addr}: {e}")
    finally:
        # This sends the final FIN packet to the client, acknowledging 
        # we are done processing.
        conn.close()
        print(f"[DISCONNECTED] {addr} disconnected.")

def start_server(host='0.0.0.0', port=65432):
    """
    Starts the main server loop.
    """
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    
    try:
        server.bind((host, port))
        server.listen(5)
        print(f"[LISTENING] Server is listening on {host}:{port}")
        
        while True:
            # Block and wait for a new connection
            conn, addr = server.accept()
            
            # Spin up a new thread for the new client so the main loop
            # can go back to waiting for the next connection immediately.
            thread = threading.Thread(target=handle_client, args=(conn, addr))
            # Daemon threads exit if the main program exits
            thread.daemon = True 
            thread.start()
            
            print(f"[ACTIVE CONNECTIONS] {threading.active_count() - 1}")
            
    except KeyboardInterrupt:
        print("\n[SHUTTING DOWN] Server stopping...")
    finally:
        server.close()

if __name__ == "__main__":
    start_server()

File: sol-go/smoke-test/sol.go
package main

import (
    "fmt"
    "io"
    "net"
    "os"
    "os/signal"
    "runtime"
    "syscall"
	"errors"
)

// handleClient handles a single client connection.
func handleClient(conn net.Conn) {
    // addr is the IP address and port of the client.
    addr := conn.RemoteAddr().String()
    fmt.Printf("[NEW CONNECTION] %s connected.\n", addr)

    // Ensure connection is closed when function exits
    defer func() {
        conn.Close()
        fmt.Printf("[DISCONNECTED] %s disconnected.\n", addr)
    }()

    buffer := make([]byte, 4096)

    for {
        // Read data from the connection
        n, err := conn.Read(buffer)

        if err != nil {
            if err == io.EOF {
                // Client shut down their sending side
                break
            }
            fmt.Printf("[ERROR] Connection error with %s: %v\n", addr, err)
            break
        }

        // Send the data back (echo)
        _, err = conn.Write(buffer[:n])
        if err != nil {
            fmt.Printf("[ERROR] Write error with %s: %v\n", addr, err)
            break
        }
    }
}

func startServer(host string, port string) {
    address := host + ":" + port
    listener, err := net.Listen("tcp", address)
    if err != nil {
        fmt.Printf("[ERROR] Could not start server: %v\n", err)
        return
    }
    // Ensure listener is closed on exit
    defer listener.Close()

    fmt.Printf("[LISTENING] Server is listening on %s\n", address)

    // Handle graceful shutdown
    c := make(chan os.Signal, 1)
    signal.Notify(c, os.Interrupt, syscall.SIGTERM)
    go func() {
        <-c // Block until signal is received
        fmt.Println("\n[SHUTTING DOWN] Server stopping...")
        listener.Close() 
    }()

    for {
        conn, err := listener.Accept()
        if err != nil {
            // Check if the error is essentially "Listener Closed"
            if errors.Is(err, net.ErrClosed) {
                // This is expected during shutdown, simply return to stop the loop
                return
            }
            
            fmt.Printf("[ERROR] Accept error: %v\n", err)
            continue
        }

        go handleClient(conn)

        // Adjust active count (approximate)
        fmt.Printf("[ACTIVE CONNECTIONS] %d\n", runtime.NumGoroutine()-2)
    }
}

func main() {
    startServer("0.0.0.0", "65432")
}
